// promptfoo/dist/src/quiverai-SJqXxYQi.js
import "./chunk-FhC4c-0y.js";
import "./logger-q5I8CByj.js";
import "./esm-D40XsWlR.js";
import "./pythonUtils-eNq6Wsfr.js";
import "./types-CyCiwPFa.js";
import "./util-DKlIKOnZ.js";
import "./fetch-4xPs0SRF.js";
import "./cache-DSW7FdwS.js";
import "./genaiTracer-CNawybn-.js";
import { t as OpenAiChatCompletionProvider } from "./chat-CaE-RGG8.js";
import "./transform-DF_Z_INN.js";
import "./openai-CQozD_Ef.js";
import "./util-ktUxLAOK.js";

//#region src/providers/quiverai.ts
const QUIVERAI_API_BASE_URL = "https://api.quiver.ai/v1";
/**
* QuiverAI chat provider extends OpenAI chat completion provider.
* QuiverAI's chat API is OpenAI-compatible and excels at SVG generation.
*/
var QuiverAiChatProvider = class extends OpenAiChatCompletionProvider {
	constructor(modelName, providerOptions = {}) {
		const config = providerOptions.config || {};
		super(modelName, {
			...providerOptions,
			config: {
				...config,
				apiBaseUrl: config.apiBaseUrl || QUIVERAI_API_BASE_URL,
				apiKeyEnvar: "QUIVERAI_API_KEY"
			}
		});
	}
	id() {
		return `quiverai:${this.modelName}`;
	}
	toString() {
		return `[QuiverAI Provider ${this.modelName}]`;
	}
	getApiUrlDefault() {
		return QUIVERAI_API_BASE_URL;
	}
};
function createQuiverAiProvider(providerPath, providerOptions = {}, env) {
	const splits = providerPath.split(":");
	const modelType = splits[1];
	const effectiveEnv = env || providerOptions.env;
	return new QuiverAiChatProvider((modelType === "chat" ? splits.slice(2) : splits.slice(1)).join(":") || "arrow-0.5", {
		...providerOptions,
		env: effectiveEnv
	});
}

//#endregion
export { createQuiverAiProvider };
